apps/src/blockly/addons/cdoConnectionChecker.js

```diff
@@ -0,0 +1,58 @@
+import GoogleBlockly from 'blockly/core';
+import {customConnectionBlockTypes} from './cdoConstants';
+
+/**
+ * A connection checker that imposes stricter typing rules for CDO custom
+ * connection type checks, but uses the same rules for dragging and safety.
+ * This checker still expects nullable arrays of string for connection
+ * type checks, and still looks for intersections in the arrays. Unlike the
+ * default checker, check arrays that contain a custom connection type are
+ * only compatible with other check arrays with the same connection type.
+ * @implements {Blockly.IConnectionChecker}
+ */
+
+export default class CdoConnectionChecker extends GoogleBlockly.ConnectionChecker {
+  /**
+   * Constructor for the connection checker.
+   */
+  constructor() {
+    super();
+  }
+
+  /**
+   * Check whether this connection is compatible with another connection with
+   * respect to the value type system.  E.g. square_root("Hello") is not
+   * compatible. Includes custom logic for CDO custom connection type checks.
+   *
+   * @param a Connection to compare.
+   * @param b Connection to compare against.
+   * @returns True if the connections share a type.
+   */
+  doTypeChecks(a, b) {
+    const checkArrayOne = a.getCheck();
+    const checkArrayTwo = b.getCheck();
+
+    // For our custom implementation, connections are rejected if only one
+    // check array contains a CDO custom connection type (Behaviors, Locations,
+    // Sprites). E.g. A block with a sprite input should not accept a number.
+    const customConnectionChecks = Object.values(customConnectionBlockTypes);
+    const checkArrayOneContainsCustomType = !!checkArrayOne?.some(check =>
+      customConnectionChecks.includes(check)
+    );
+    const checkArrayTwoContainsCustomType = !!checkArrayTwo?.some(check =>
+      customConnectionChecks.includes(check)
+    );
+    if (checkArrayOneContainsCustomType !== checkArrayTwoContainsCustomType) {
+      return false;
+    }
+
+    // Based on:
```

Comment:
Could we just call on `super.doTypeChecks` here?

Reply:
I tried replacing everything below with `super.doTypeCheck(a,b)` and it breaks things. Perhaps I don't know enough about how extended classes work, but it seems like it's simpler to just to leave the full logic here.

Reply:
Interesting! I wonder why it's not working. Sorry about that! I'll do a little digging bc I'm curious.

Reply:
This worked! Seems obvious in hindsight. Thanks!